### DISCUSSION TITLE
Navigation works only once in my App

### DISCUSSION THREAD

**Discussion by pkarsten** [unoplatform/uno#17361] [Q&A] [answered]
I don't know if I have understood the navigation with the Navigator correctly. 

in App.xaml.cs i have: 

```
    private static void RegisterRoutes(IViewRegistry views, IRouteRegistry routes)
    {
        views.Register(
            new ViewMap(ViewModel: typeof(ShellViewModel)),
            new ViewMap<MainPage, MainViewModel>(),
            new ViewMap<SettingsPage, SettingsViewModel>(),
            new ViewMap<KioskPage, KioskViewModel>(),
            new DataViewMap<SecondPage, SecondViewModel, Entity>()
        );

        routes.Register(
            new RouteMap("", View: views.FindByViewModel<ShellViewModel>(),
                Nested:
                [
                    new ("Main", View: views.FindByViewModel<MainViewModel>()),
                    new ("Second", View: views.FindByViewModel<SecondViewModel>()),
                    new ("Settings", View: views.FindByViewModel<SettingsViewModel>()),
                    new ("Kiosk", View: views.FindByViewModel<KioskViewModel>())
                ]
            )
        );
    }
```

in ShellViewModel.cs i have: 

```
   public async Task Start()
    {
        await _navigator.NavigateViewModelAsync<KioskViewModel>(this);
    }

```
My App starts and goes to KioskPage.xaml

KioskPage.xaml.cs looks like: 

```
public sealed partial class KioskPage : Page
{
    public KioskPage()
    {
        this.InitializeComponent();
    }
}
```

i ihave a Model KioskViewModel: 

```
  public KioskViewModel(INavigator navigator, IMyService myservice)
    {
        _navigator = navigator;
        MessageService = myservice;
        GoToSettings = new AsyncRelayCommand(OnNavigateToSettingsPage);
    }
    private async Task OnNavigateToSettingsPage()
    {
        await StopClock();
        await _navigator.NavigateViewModelAsync<SettingsViewModel>(this);
    }
```

then i have SettingsViewModel.cs : 
```
    public SettingsViewModel(INavigator navigator, IMyService myservice, LoadImagesBackgroundworker backgroundWorkerService)
    {
        _navigator = navigator;
        MyService = myservice;
        LoadPicsCommand = new AsyncRelayCommand(OnLoadPics);
        Submit = new AsyncRelayCommand(OnSaveSettings);
        LogInLogOut = new AsyncRelayCommand(OnLoginLogout);
        GoToDashBoardPage = new AsyncRelayCommand(OnNavigateToDashBoardPage);
}
```

```
private async Task OnNavigateToDashBoardPage()
    {
        await _navigator.NavigateViewModelAsync<KioskViewModel>(this);
    }
```

The app starts and KioskPage is called up, I click on Settings and get to the SettingsPage. If I want to go back to the KioskPage on the SettingsPage, it doesn't work, nothing happens when I click on the button. 

What I want: my app should start, call up one or the other page (Settings or Kiosk) depending on a property, and I always have to switch back and forth between KioskPage and SettingsPage. 

What am I doing wrong? How can I do it right? 

Thank you very much

**Comment by rajamatt** [comment, marked as answer]
Hi @pkarsten,

I had a hard time figuring out what was causing the problem by looking at your code as everything navigation seemed fine. I ended up finding out that it was the starting out on KioskModel, then the NavigateViewModelAsync<SettingsModel> into the NavigateViewModelAsync<KioskModel> that was causing the issue. The problem being that the navigation back stack is causing the last navigation to the KioskPage to display the SettingsPage instead.

As a workaround, I can suggest clearing the backstack when navigating to the KioskPage from the SettingsModel:

```csharp
private async Task OnNavigateToDashBoardPage()
{
    await _navigator.NavigateViewModelAsync<KioskViewModel>(this, qualifiers: Qualifiers.ClearBackStack);
}
```

If what you actually want is back navigation, that would work too:

```csharp
private async Task OnNavigateToDashBoardPage()
{
    await _navigator.NavigateBackAsync(this);
}
```

For a scenario like yours where you'd want to switch between the Settings/Kiosk page continuously, I would maybe suggest using a `NavigationView` as seen here in [Navigating to page elements](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/Advanced/HowTo-UseNavigationView.html#navigating-to-page-elements). It requires some Region attaching but it would offer the same experience without having to use navigation explicitly in the code behind. You can see it in action (sometimes as `TabBar` instead of `NavigationView`) in [Commerce](https://github.com/unoplatform/Uno.Samples/tree/master/reference/Commerce) and [ToDo](https://github.com/unoplatform/Uno.Samples/tree/master/reference/ToDo).

**Reply by pkarsten** [reply]
@rajamatt Thank you
